server/server.go

```diff
@@ -42,72 +55,362 @@ func HandleWebsocketConnection(serverHTTPResponse http.ResponseWriter, clientHTT
 	// Read the response from the client which should contain his IdenetityPublicKey and the signed byte sequence
-	_, clientMessage, readMessageErr := websocketConnection.ReadMessage()
+	_, messageFromClientProto, readMessageErr := websocketConnection.ReadMessage()
 	if readMessageErr != nil {
 		// If there is an error while reading the response from the client, terminate the connection
-		return websocket.BinaryMessage, readMessageErr
+		return "", readMessageErr
 	}
 	// Unmarshal the response from the client into our protobuf Auth structure
-	protoUnmarshalErr := proto.Unmarshal(clientMessage, &protobufAuth)
+	protoUnmarshalErr := proto.Unmarshal(messageFromClientProto, &messageFromClient)
 	if protoUnmarshalErr != nil {
 		// If there is an error while unmarshaling the response from the client, terminate the connection
-		return websocket.BinaryMessage, protoUnmarshalErr
+		return "", protoUnmarshalErr
 	}
 	// Create a [32]byte{} identityPublicKey to satisfy ed25519.Verify() type requirements
 	identityPublicKey := [32]byte{}
 	// Create a [64]byte{} signedByteSequenceFromClient to satisfy ed25519.Verify() type requirements
 	signedByteSequenceFromClient := [64]byte{}
+	// Create a string representation of the Identity Public Key
+	identityPublicKeyString := string(messageFromClient.Response.Auth.IdentityPublicKey)
 	// Decode the hex representation of the identityPublicKey here as transmitting it over the network without hex encoding can cause issues with some clients
-	identityPublicKeyDecoded, identityPublicKeyDecodedErr := hex.DecodeString(string(protobufAuth.GetIdentityPublicKey()))
+	identityPublicKeyDecoded, identityPublicKeyDecodedErr := hex.DecodeString(identityPublicKeyString)
 	if identityPublicKeyDecodedErr != nil {
 		// If there is an error while decoding the hex identity public key, terminate the connection
-		return websocket.BinaryMessage, identityPublicKeyDecodedErr
+		return "", identityPublicKeyDecodedErr
+	}
+	// Get the byte sequence which was signed by the client
+	byteSequenceThatClientSigned := messageFromClient.Response.Auth.ToSign
+	// Check if the byte sequence that was signed by the client contains the initial bytes we sent to the client
+	if !bytes.HasPrefix(byteSequenceThatClientSigned, backendRandomBytes) {
+		// If the client has modified the bytes we sent, return an error pointing out that this behavior is not allowed
+		return "", errors.New("Client is only allowed to append a byte sequence, and not to modify the one which was sent")
 	}
 	// Fill the newly created identityPublicKey with the hex decoded representation of the IdentityPublicKey contained in the response from the client
 	copy(identityPublicKey[:], identityPublicKeyDecoded)
 	// Fill the newly created signedByteSequenceFromClient with the Signature contained in the response from the client
-	copy(signedByteSequenceFromClient[:], protobufAuth.GetSignature())
+	copy(signedByteSequenceFromClient[:], messageFromClient.Response.Auth.Signature)
```

Comment:
It might make sense to validate the length of the signature as well as the length of the id pub key. Just to be sure. Once I had a problem were I copied over too less bytes on an invalid request. End of the story, I had to debug for 2 hours. By checking if the id pub key is 32 bytes long we might be able to make debugging easier in the case we have to.

Reply:
Agreed, better be safe than sorry. Will add the required checks.